### PR TITLE
oops_concepts and block procs and lambda understanding

### PR DESCRIPTION
**Pull request by mayuriv-webonise**

### CODE REVIEW

File: assignment2.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+service_project = Project.new('Service', 100000, 50000)
+product_project = Project.new('Product', 500000, 200000)
+
+puts "Engineering Department Employees: #{engineering.get_department_employees}"
+puts "Service Project Revenue: #{service_project.revenue}"
+puts "Product Project Profit: #{product_project.profit}"
+
+  
+
```
Comment (venkatesh-webonise) [inline assignment2.rb]: Please fix this new line error and remove empty lines

File: assignment2.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    attr_accessor :type, :revenue, :profit
+  
+    def initialize(type, revenue, profit)
+      @type = type
+      @revenue = revenue
+      @profit = profit
+    end
+  end
+
+  hr = Hr.new
```
Comment (venkatesh-webonise) [inline assignment2.rb]: I don't see `Company` class at all in this code. Please check each and every point of the assignment again.

File: assignment2.rb (hunk 1 above)
Comment (mayuriv-webonise) [reply assignment2.rb]: removed new lines

File: assignment2.rb (hunk 2 above)
Comment (mayuriv-webonise) [reply assignment2.rb]: Added company class and worked on the points
Only the HR can add/edit details of the employee.

Only the HR can get the details of all the employees. 

The department heads can get the details of those employees in his department.

Projects in the company can either be a Service or Product

Each service offered by the company has 1 external client 

Each product developed by the company has several clients. 

The company keeps track of the yearly revenue and profit for the projects. The profit for service & product is calculated differently.

File: assignment2.rb [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  def add_employee(hr, employee)
+    hr.add_employee(employee)
+  end
+
+  def get_all_employee_details(hr)
+    hr.get_all_employee_details
+  end
+  def calculate_yearly_revenue()
+    
+    @projects.each do |project|
```
Comment (venkatesh-webonise) [inline assignment2.rb:134]: please check the alignment here.

File: assignment2.rb [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    end
+  end
+  def calculate_yearly_profit()
+  
+    @projects.each do |project|
+     if project.type == 'Service'
+
+         @service_profit =  @service_profit + project.profit
+     else
+
```
Comment (venkatesh-webonise) [inline assignment2.rb]: same here

File: assignment2.rb [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  def add_department(department)
+    @departments << department
+  end
+
+  def add_project(project)
+    @projects << project
+  end
+
+  def add_employee(hr, employee)
```
Comment (venkatesh-webonise) [inline assignment2.rb]: Company class shouldn't add employees. Please remove this method.

File: assignment2.rb [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+engineering_department = Department.new('Engineering', 'Engineering Department Head')
+sales_department = Department.new('Sales', 'Sales Department Head')
+marketing_department = Department.new('Marketing', 'Marketing Department Head')
+
+company.add_department(hr_department)
+company.add_department(engineering_department)
+company.add_department(sales_department)
+company.add_department(marketing_department)
+
+service_project = Project.new('Healthcare web','Service', 'client 1', 50000)
```
Comment (venkatesh-webonise) [inline assignment2.rb]: no need to have type in the `Project` class as we only have 2 types. You can directly create `ServiceProject` and `ProductProject` something like this.

File: assignment2.rb (hunk 3 above)
Comment (mayuriv-webonise) [reply assignment2.rb:134]: corrected

File: assignment2.rb (hunk 4 above)
Comment (mayuriv-webonise) [reply assignment2.rb]: corrected

File: assignment2.rb (hunk 5 above)
Comment (mayuriv-webonise) [reply assignment2.rb]: Removed

**Review by venkatesh-webonise — APPROVED** [review]

**Pull request merged**